lib/src/buffer.dart

```diff
@@ -36,109 +30,95 @@ class _Buffer<T> implements StreamTransformer<T, List<T>> {
 
   @override
   Stream<List<T>> bind(Stream<T> values) {
-    StreamController<List<T>> controller;
-    if (values.isBroadcast) {
-      controller = new StreamController<List<T>>.broadcast();
-    } else {
-      controller = new StreamController<List<T>>();
-    }
+    var controller = values.isBroadcast
+        ? new StreamController<List<T>>.broadcast(sync: true)
+        : new StreamController<List<T>>(sync: true);
 
     List<T> currentResults;
     bool waitingForTrigger = true;
-    StreamSubscription valuesSub;
+    bool isTriggerDone = false;
+    bool isValueDone = false;
+    StreamSubscription valueSub;
     StreamSubscription triggerSub;
 
-    cancelValues() {
-      var sub = valuesSub;
-      valuesSub = null;
-      return sub?.cancel() ?? new Future.value();
-    }
-
-    cancelTrigger() {
-      var sub = triggerSub;
-      triggerSub = null;
-      return sub?.cancel() ?? new Future.value();
-    }
-
-    closeController() {
-      var ctl = controller;
-      controller = null;
-      return ctl?.close() ?? new Future.value();
-    }
-
     emit() {
       controller.add(currentResults);
       currentResults = null;
       waitingForTrigger = true;
     }
 
     onValue(T value) {
-      currentResults = _collectToList(value, currentResults);
-      if (!waitingForTrigger) {
-        emit();
+      (currentResults ??= <T>[]).add(value);
+
+      if (!waitingForTrigger) emit();
+
+      if (isTriggerDone) {
+        valueSub.cancel();
+        controller.close();
       }
     }
 
-    valuesDone() {
-      valuesSub = null;
+    onValuesDone() {
+      isValueDone = true;
       if (currentResults == null) {
-        closeController();
-        cancelTrigger();
+        triggerSub?.cancel();
+        controller.close();
       }
     }
 
     onTrigger(_) {
-      if (currentResults == null) {
-        waitingForTrigger = false;
-        return;
-      }
-      emit();
-      if (valuesSub == null) {
-        closeController();
-        cancelTrigger();
+      waitingForTrigger = false;
+
+      if (currentResults != null) emit();
+
+      if (isValueDone) {
+        triggerSub.cancel();
```

Comment:
If it was null we wouldn't be getting any events through it, so here yes I think it is guaranteed.